### PR TITLE
fix: mapping prune

### PR DESCRIPTION
**Pull request by krackerman**

### CODE REVIEW

**Comment by krackerman** [comment]
https://github.com/loopmediagroup/service-search/issues/1910

File: README.md [hunk 1]
```diff
@@ -380,6 +380,7 @@ Interacting with the rest api of Elasticsearch
 - `mapping.get(name: String)` - get mapping details from Elasticsearch (current version)
 - `mapping.historic(name: String)` - get _old_ mapping versions and their respective document counts from Elasticsearch
 - `mapping.list()` - Lists all mappings currently in Elasticsearch
+- `mapping.prune(index: String)` - Removes indices from Elasticsearch that are not persisted
```
Comment (simlu) [inline README.md:383]:
```suggestion
- `mapping.prune(index: String)` - Removes index versions from Elasticsearch that are not tracked
```

File: src/util/rest/mapping/prune.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+const assert = require('assert');
+
+const getIndices = async (call, idx) => {
```
Comment (simlu) [inline src/util/rest/mapping/prune.js:3]:
```suggestion
const getIndexVersions = async (call, idx) => {
```

File: src/util/rest/mapping/prune.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  const result = await call('GET', '', {
+    endpoint: `_cat/indices/${idx}@*`
+  });
+  return result.body.map(({ index }) => index);
+};
+
+const deleteIndex = (call, idx) => call('DELETE', idx)
```
Comment (simlu) [inline src/util/rest/mapping/prune.js:10]:
```suggestion
const deleteIndexVersion = (call, idx, version) => call('DELETE', `${idx}@${version}`)
```

File: src/util/rest/mapping/prune.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  .then((r) => r.statusCode === 200 && r.body.acknowledged === true);
+
+module.exports = async (call, versions, idx) => {
+  const indices = await getIndices(call, idx);
```
Comment (simlu) [inline src/util/rest/mapping/prune.js:14]:
```suggestion
  const removeVersions = await getIndices(call, idx);
```

File: src/util/rest/mapping/prune.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  const indexVersions = versions.get(idx);
```
Comment (simlu) [inline src/util/rest/mapping/prune.js:15]:
```suggestion
  const localeVersions = versions.get(idx);
```

File: src/util/rest/mapping/prune.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  assert(indexVersions !== undefined, 'Index must be loaded');
```
Comment (simlu) [inline src/util/rest/mapping/prune.js:16]:
```suggestion
  assert(indexVersions !== undefined, 'Index must known and loaded');
```

File: src/util/rest/mapping/prune.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    endpoint: `_cat/indices/${idx}@*`
+  });
+  return result.body.map(({ index }) => index.split('@')[1]);
+};
+
+const deleteIndexVersion = (call, idx, version) => call('DELETE', `${idx}@${version}`)
+  .then((r) => r.statusCode === 200 && r.body.acknowledged === true);
+
+module.exports = async (call, versions, idx) => {
+  const remoteVersions = await getIndexVersions(call, idx);
```
Comment (simlu) [inline src/util/rest/mapping/prune.js:14]: move this below the cheap call

File: src/util/rest/mapping/prune.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  const localVersions = versions.get(idx);
+  assert(localVersions !== undefined, 'Index must be loaded');
+  const persistedVersions = Object.keys(localVersions);
```
Comment (simlu) [inline src/util/rest/mapping/prune.js:17]:
```suggestion
  const localVersions = Object.keys(versions.get(idx));
```
and raise in versions.get

**Review by simlu — APPROVED** [review]

**Review by simlu — APPROVED** [review]

**Pull request merged**